siwe_test.go

```diff
@@ -0,0 +1,345 @@
+	}
+}
+
+func assertCase(t *testing.T, fields map[string]interface{}, parsed string, key string) {
+	if field, ok := fields[key]; ok {
+		assert.Equal(t, field, parsed, "%s should be %s", key, field)
+	}
+}
+
+func parsingNegative(t *testing.T, cases map[string]interface{}) {
+	for name, message := range cases {
+		_, err := ParseMessage(message.(string))
+		assert.Error(t, err, name)
+	}
+}
+
+func parsingPositive(t *testing.T, cases map[string]interface{}) {
+	for name, v := range cases {
+		data := v.(map[string]interface{})
+		message := data["message"].(string)
+		fields := data["fields"].(map[string]interface{})
+		parsed, err := parseMessage(message)
+		assert.Nil(t, err, name)
+
+		assertCase(t, fields, parsed["domain"].(string), "domain")
+		assertCase(t, fields, parsed["address"].(string), "address")
+		assertCase(t, fields, parsed["uri"].(string), "uri")
+		assertCase(t, fields, parsed["version"].(string), "version")
+		assertCase(t, fields, parsed["chainId"].(string), "chainId")
+		assertCase(t, fields, parsed["issuedAt"].(string), "issuedAt")
+
+		if val, ok := parsed["statement"]; ok {
+			assertCase(t, fields, val.(string), "statement")
+		}
+
+		if val, ok := parsed["nonce"]; ok {
+			assertCase(t, fields, val.(string), "nonce")
+		}
+	}
+}
+
+func validationNegative(t *testing.T, cases map[string]interface{}) {
+	for name, v := range cases {
+		data := v.(map[string]interface{})
+		message, err := InitMessage(
+			data["domain"].(string),
+			data["address"].(string),
+			data["uri"].(string),
+			data["version"].(string),
+			data,
+		)
+		assert.Nil(t, err)
+
+		_, err = message.Verify(data["signature"].(string))
+
+		assert.Error(t, err, name)
+	}
+}
+
+func validationPositive(t *testing.T, cases map[string]interface{}) {
+	for name, v := range cases {
+		data := v.(map[string]interface{})
+		message, err := InitMessage(
+			data["domain"].(string),
+			data["address"].(string),
+			data["uri"].(string),
+			data["version"].(string),
+			data,
+		)
+		assert.Nil(t, err)
+
+		_, err = message.Verify(data["signature"].(string))
+
+		assert.Nil(t, err, name)
+	}
+}
+
+func TestGlobalTestVector(t *testing.T) {
+	files := make(map[string]*os.File, 4)
+	for test, filename := range map[string]string{
+		"parsing-negative":    "siwe-js/test/parsing_negative.json",
+		"parsing-positive":    "siwe-js/test/parsing_positive.json",
+		"validation-negative": "siwe-js/test/validation_negative.json",
+		"validation-positive": "siwe-js/test/validation_positive.json",
+	} {
+		file, err := os.Open(filename)
+		assert.Nil(t, err)
+		files[test] = file
+		defer file.Close()
+	}
+
+	for test, file := range files {
+		data, _ := ioutil.ReadAll(file)
+
+		var result map[string]interface{}
+		err := json.Unmarshal([]byte(data), &result)
+		assert.Nil(t, err)
+
+		switch test {
+		case "parsing-negative":
+			parsingNegative(t, result)
+		case "parsing-positive":
+			parsingPositive(t, result)
```

Comment:
You can use those tests to test the serialisation as well, by checking the round trip eip4361 -> object -> eip4361